File: components/engine/src/test/java/org/storm/engine/example/CircleCornerTestState.java
package org.storm.engine.example;

import org.storm.core.input.action.ActionManager;
import org.storm.engine.KeyActionConstants;
import org.storm.engine.request.RequestQueue;
import org.storm.engine.request.types.TogglePhysicsRequest;
import org.storm.engine.request.types.VelocityUpdateRequest;
import org.storm.physics.constants.Vectors;
import org.storm.physics.entity.Entity;
import org.storm.physics.math.Vector;
import org.storm.physics.math.geometry.shapes.Circle;
import org.storm.sound.types.MediaSound;

public class CircleCornerTestState extends SwitchableState {

  private final Entity player = new EntityImpl(
    new Circle(
      this.unitConvertor.toUnits(25),
      this.unitConvertor.toUnits(200),
      this.unitConvertor.toUnits(10)
    ),
    this.unitConvertor.toUnits(2),
    10,
    0.1
  );

  public CircleCornerTestState() {
    super();
    this.entities.add(new ImmovableRectEntity(this.unitConvertor.toUnits(200), this.unitConvertor.toUnits(200), this.unitConvertor.toUnits(100), this.unitConvertor.toUnits(100)));
    this.entities.add(player);
    this.soundManager.add("bgm", new MediaSound("src/test/resources/music/bgm.mp3", true));
    this.soundManager.adjustAllVolume(0.1);
  }

  @Override
  public void load(RequestQueue requestQueue) {
    this.soundManager.play("bgm");
    requestQueue.submit(new TogglePhysicsRequest(false));
  }

  @Override
  public void unload(RequestQueue requestQueue) {
    this.soundManager.pause("bgm");
  }

  @Override
  public void process(ActionManager actionManager, RequestQueue requestQueue) {
    super.process(actionManager, requestQueue);

    Vector velocity = Vectors.ZERO_VECTOR;

    if (actionManager.isPerforming(KeyActionConstants.W)) {
      velocity = velocity.add(Vectors.UNIT_NORTH.scale(this.player.getSpeed()));
    }
    else if (actionManager.isPerforming(KeyActionConstants.S)) {
      velocity = velocity.add(Vectors.UNIT_SOUTH.scale(this.player.getSpeed()));
    }
    else if (actionManager.isPerforming(KeyActionConstants.D)) {
      velocity = velocity.add(Vectors.UNIT_EAST.scale(this.player.getSpeed()));
    }
    else if (actionManager.isPerforming(KeyActionConstants.A)) {
      velocity = velocity.add(Vectors.UNIT_WEST.scale(this.player.getSpeed()));
    }

    requestQueue.submit(new VelocityUpdateRequest(player, velocity));
  }
}

File: components/engine/src/test/kotlin/org/storm/engine/example/CircleCornerTestState.kt
package org.storm.engine.example

import org.storm.core.context.Context
import org.storm.core.context.YAML_MAPPER
import org.storm.core.event.EventManager
import org.storm.core.extensions.units
import org.storm.core.graphics.canvas.Canvas
import org.storm.core.input.ActionState
import org.storm.core.sound.Sound
import org.storm.core.sound.SoundManager
import org.storm.engine.Controls
import org.storm.engine.events.EngineEvent
import org.storm.engine.events.getEngineEventStream
import org.storm.physics.PhysicsEngine
import org.storm.physics.collision.Collider
import org.storm.physics.math.Vector
import org.storm.physics.math.geometry.shapes.Circle

class CircleCornerTestState : SwitchableState() {

    private data class Player(var speed: Double) {
        val collider = Collider(
            Circle(
                25.0.units,
                200.0.units,
                10.0.units
            ),
            100.0.units,
            0.1,
            this
        )
    }

    private val player = Player(2.0.units)
    private val rect = ImmovableRect(
        200.0.units,
        200.0.units,
        100.0.units,
        100.0.units
    )

    private val movementVectors = mapOf(
        Controls.W to Vector.UNIT_NORTH,
        Controls.S to Vector.UNIT_SOUTH,
        Controls.A to Vector.UNIT_WEST,
        Controls.D to Vector.UNIT_EAST,
    )

    override val colliders: Set<Collider> = setOf(
        rect.collider,
        player.collider
    )

    override suspend fun onRegister(physicsEngine: PhysicsEngine, soundManager: SoundManager) {
        val bgm = Context.YAML_MAPPER.readValue(this::class.java.getResourceAsStream("/sound/bgm.yml"), Sound::class.java)
        soundManager.add("bgm", bgm)
        soundManager.adjustAllVolume(0.1)
    }

    override suspend fun onSwapOff(physicsEngine: PhysicsEngine, soundManager: SoundManager) {
        soundManager.stop("bgm")
    }

    override suspend fun onSwapOn(physicsEngine: PhysicsEngine, soundManager: SoundManager) {
        soundManager.play("bgm")
        EventManager.getEngineEventStream().produce(EngineEvent.TogglePhysics())
    }

    override suspend fun process(actionState: ActionState) {
        super.process(actionState)

        if (actionState.isFirstActivation(Controls.SPACE)) {
            EventManager.getEngineEventStream().produce(EngineEvent.TogglePhysics())
        }

        // TODO: this is kinda jank, should update eventually
        this.player.collider.velocity = Vector.ZERO_VECTOR
        this.movementVectors.map { (action, vector) ->
            if (actionState.isActionHeld(action, 1)) {
                this.player.collider.velocity += vector.scale(this.player.speed)
            }
        }
    }

    override suspend fun render(canvas: Canvas, x: Double, y: Double) {
        rect.render(canvas, x, y)
        player.collider.render(canvas, x, y)
    }

}
